berkdb/db/db_meta.c

```diff
@@ -1023,3 +1023,80 @@ __berkdb_count_freepages(int fd)
 		free(buf);
 	return npages;
 }
+
+/*
+ * __db_estimate_fill --
+ * Look at some random pages and estimate free pages and fill factor.
+ *
+ * PUBLIC: int __db_estimate_fill __P((DB *, int, double *, double *));
+ */
+int
+__db_estimate_fill(dbp, npages, percent_freelist, avg_percent_empty)
+        DB *dbp;
+        int npages;
+        double *percent_freelist;
+        double *avg_percent_empty;
+{
+    // take 1 - just pull the pages into the buffer pool and look at them
+    int ret;
+    int nfree = 0;
+    u_int64_t freespace = 0;
+    db_pgno_t maxpages;
+    DBC *dbc = NULL;
+    DB_LOCK lk;
+    DB_MPOOLFILE *mpf = dbp->mpf;
+    PAGE *pg;
+
+    ret = __db_get_numpages(dbp, &maxpages);
+    if (ret)
+        return ret;
+
+    ret = dbp->cursor(dbp, NULL, &dbc, 0);
+    if (ret)
+        return ret;
+
+    for (int i = 0; i < npages; i++) {
+        // __db_lget(dbc, action, pgno, mode, lkflags, lockp)
+
+        db_pgno_t pgno = 1 + rand() % (maxpages-1);
+        ret = __db_lget(dbc, LCK_ALWAYS, pgno, DB_LOCK_READ, 0, &lk);
+        if (ret)
+            goto done;
+
+        if ((ret = __memp_fget(mpf, &pgno, 0, &pg)) != 0) {
+            (void)__TLPUT(dbc, lk);
+            goto done;
+        }
+
+        if (TYPE(pg) == P_INVALID) {
+            nfree++;
+        }
+        else {
+            int fs = P_FREESPACE(dbp, pg);
+            freespace += P_FREESPACE(dbp, pg);
+        }
+
+        ret = __memp_fput(mpf, pg, 0);
+        if (ret) {
+            (void)__TLPUT(dbc, lk);
+            goto done;
+        }
+        (void)__TLPUT(dbc, lk);
+    }
+
+    u_int32_t pgsize;
+    ret = dbp->get_pagesize(dbp, &pgsize);
+    if (ret)
+        goto done;
+    *percent_freelist = ((double) nfree /  npages) * 100;
+    *avg_percent_empty = 100 - (((double) freespace / (npages * pgsize)) * 100);
```

Comment:
Sorry for nitpicking but I think it's better to subtract `SIZEOF_PAGE` (the overhead of `PAGE` itself) from `pgsize`.